test/IDE/complete_ambiguous.swift

```diff
@@ -473,3 +473,88 @@ func testBestSolutionGeneric() {
 // BEST_SOLUTION_FILTER_GEN-DAG: Keyword[self]/CurrNominal:     self[#Test1#]; name=self
 // BEST_SOLUTION_FILTER_GEN: End completions
 
+func testAmbiguousArgs() {
+  struct A {
+    func someFunc(a: Int, b: Int) -> A { return self }
+    func variadic(y: Int, x: Int...)
+  }
+
+  struct B {
+    func someFunc(a: Int, c: String = "", d: String = "") {}
+  }
+
+  func overloaded() -> A { return A() }
+  func overloaded() -> B { return B() }
+
+  let localInt = 10
+  let localString = "Hello"
+
+  overloaded().someFunc(a: 2, #^ARG_NO_LABEL^#)
+
+  // ARG_NO_LABEL: Begin completions, 3 items
+  // ARG_NO_LABEL-DAG: Pattern/Local/Flair[ArgLabels]:     {#b: Int#}[#Int#]; name=b: Int
+  // ARG_NO_LABEL-DAG: Pattern/Local/Flair[ArgLabels]:     {#c: String#}[#String#]; name=c: String
+  // ARG_NO_LABEL-DAG: Pattern/Local/Flair[ArgLabels]:     {#d: String#}[#String#]; name=d: String
+  // ARG_NO_LABEL: End completions
+
+  overloaded().someFunc(a: 2, b: #^ARG_LABEL^#)
+
+  // ARG_LABEL: Begin completions
+  // ARG_LABEL-DAG: Decl[LocalVar]/Local:                          localString[#String#]; name=localString
+  // ARG_LABEL-DAG: Decl[LocalVar]/Local/TypeRelation[Identical]:  localInt[#Int#]; name=localInt
+  // ARG_LABEL: End completions
+
+  overloaded().someFunc(a: 2, c: "Foo", #^ARG_NO_LABEL2^#)
+
+  // ARG_NO_LABEL2: Begin completions, 1 item
+  // ARG_NO_LABEL2: Pattern/Local/Flair[ArgLabels]:     {#d: String#}[#String#]; name=d: String
+  // ARG_NO_LABEL2: End completions
+
+  overloaded().someFunc(a: 2, wrongLabel: "Foo", #^ARG_NO_LABEL_PREV_WRONG^#)
+
+  // ARG_NO_LABEL_PREV_WRONG: Begin completions, 2 items
+  // ARG_NO_LABEL_PREV_WRONG-DAG: Pattern/Local/Flair[ArgLabels]:     {#c: String#}[#String#]; name=c: String
+  // ARG_NO_LABEL_PREV_WRONG-DAG: Pattern/Local/Flair[ArgLabels]:     {#d: String#}[#String#]; name=d: String
```

Comment:
Yeah this seems wrong looking at it again. It's not matching positionally, the b overload is considered worse in the scoring because it fails to find an arg for b after a (and b is required) and then reports wrongLabel and the completion as extraneous arguments. The c:d: overload has nothing missing because c and d are defaulted, so just sees wrongLabel and the completion as extraneous and gets a better score. We don't actually bind any parameter to the completion in either case so it hits my fallback logic (needed for the existing `OVERLOAD_LABEL2` case in test/IDE/completion_call_arg.swift) that binds it to the first unfulfilled parameter after the last fulfilled one (if there is one). I think there's a few things to try:
1) Update the matching logic to somehow skip over the bad label when binding, so there's no missing param, the completion arg is bound in both overloads, and they both just have one extraneous argument (and end up with the same score). This might have some undesirable consequences elsewhere though.
2) Just give up when we don't get a binding for the completion, giving no results in this case and the existing `OVERLOAD_LABEL2` case (regressing it, but perhaps understandably?).
3) Change the filtering logic to keep all invalid solutions when there are no valid ones, rather than just the best one. From memory that caused some tests to fail, mostly around completion after == because it has tons of overloads and the fix to allow type mismatches made them all viable regardless of the type on the LHS. I suspect it might also give us some nonsense results for solutions with tons of fixes.

Reply:
Got option 1 working without any other impact.